Sources/NIOHTTP1/HTTPServerUpgradeHandler.swift

```diff
@@ -186,12 +186,12 @@ public final class HTTPServerUpgradeHandler: ChannelInboundHandler, RemovableCha
             return context.eventLoop.makeSucceededFuture(nil)
         }
 
-        guard let upgrader = upgraders[proto.lowercased()] else {
+        guard let upgrader = self.upgraders[proto.lowercased()] else {
             return self.handleUpgradeForProtocol(context: context, protocolIterator: protocolIterator, request: request, allHeaderNames: allHeaderNames, connectionHeader: connectionHeader)
         }
 
         let requiredHeaders = Set(upgrader.requiredUpgradeHeaders.map { $0.lowercased() })
-        guard requiredHeaders.isSubset(of: allHeaderNames) && requiredHeaders.isSubset(of: connectionHeader) else {
+        guard requiredHeaders.isSubset(of: allHeaderNames) && connectionHeader.isSubset(of: allHeaderNames) else {
```

Comment:
I don't think the original check was incorrect. An `Upgrade` header is by definition a hop-by-hop header field, and so must be mentioned in the `Connection` header field value. RFC 7230 § 6.7 calls this out explicitly:

> When Upgrade is sent, the sender MUST also send a Connection header field (Section 6.1) that contains an "upgrade" connection option, in order to prevent Upgrade from being accidentally forwarded by intermediaries that might not implement the listed protocols

Somewhat by definition, then, any header required for upgrade must *also* be a hop-by-hop header, otherwise it will be sent unnecessarily when passing through intermediaries that do not support the upgrade. For this reason, we validate that.

In the case of protocols that have implementations that incorrectly fail to do this (most notably WebSocket), this can be handled by implementing an empty required header array and then checking in the upgrade callback.

Reply:
I think I slightly misunderstood the spec. Thanks for the description. You are right, the original code is correct. I will set this back.